src/main/java/org/broadinstitute/hellbender/utils/downsampling/PositionalDownsampler.java

```diff
@@ -71,16 +73,31 @@ private void handlePositionalChange( final GATKRead newRead ) {
         // Use ReadCoordinateComparator to determine whether we've moved to a new start position.
         // ReadCoordinateComparator will correctly distinguish between purely unmapped reads and unmapped reads that
         // are assigned a nominal position.
-        if ( previousRead != null && ReadCoordinateComparator.compareCoordinates(previousRead, newRead, header) != 0 ) {
-            if ( reservoir.hasFinalizedItems() ) {
-                finalizeReservoir();
+        if ( previousRead != null) {
+            final int cmpDiff = ReadCoordinateComparator.compareCoordinates(previousRead, newRead, header);
+            if (cmpDiff == 1) {
+                throw new GATKException.ShouldNeverReachHereException(
+                        String.format("Reads must be coordinate sorted (earlier %s later %s)", previousRead, newRead));
+            }
+            if (cmpDiff != 0) {
+                reservoir.signalEndOfInput();
+                if ( reservoir.hasFinalizedItems() ) {
+                    finalizeReservoir();
+                }
             }
         }
     }
 
     private void finalizeReservoir() {
+        // We can't consume finalized reads from the reservoir unless we first signal EOI.
+        // Once signalEndOfInput has been called and propagated to the ReservoirDownsampler, consumeFinalizedItems
+        // must be called on the ReservoirDownsampler before any new items can be submitted to it, to reset it's
+        // state so it can be recycled/reused for the next downsampling position.
+        reservoir.signalEndOfInput();
         finalizedReads.addAll(reservoir.consumeFinalizedItems());
+        reservoir.clearItems();
         reservoir.resetStats();
+        previousRead = null;
```

Comment:
Is it necessary to set `previousRead` to null here? Does this fix an actual problem?

Reply:
Yes, see my response comment above.